File: apps/web/src/lib/api/memory.ts
import { apiRequest } from "@/lib/api/client";

export type MemoryKind =
  | "career_goal"
  | "conversation"
  | "conversation_summary"
  | "favorite_verse"
  | "journal"
  | "reflection"
  | "note";

export interface MemoryItem {
  id: string;
  kind: MemoryKind;
  title: string | null;
  content: string;
  source_id: string | null;
  metadata: Record<string, unknown>;
  created_at: string | null;
  updated_at: string | null;
}

export interface CareerGoal {
  id: string;
  title: string;
  description: string;
  status: string;
  memory_item_id: string | null;
  created_at: string;
  updated_at: string;
}

export interface FavoriteVerse {
  id: string;
  verse_id: string;
  note: string | null;
  memory_item_id: string | null;
  citation: string | null;
  translation: string | null;
  created_at: string;
}

export interface JournalEntry {
  id: string;
  title: string | null;
  content: string;
  mood_note: string | null;
  created_at: string;
  updated_at: string;
}

export interface ReflectionEntry {
  id: string;
  journal_id: string;
  verse_id: string | null;
  content: string;
  created_at: string;
}

export async function listMemoryItems(kind?: MemoryKind): Promise<MemoryItem[]> {
  const qs = kind ? `?kind=${kind}` : "";
  return apiRequest<MemoryItem[]>(`/memory/items${qs}`, { auth: true });
}

export async function createMemoryItem(input: {
  kind?: MemoryKind;
  title?: string;
  content: string;
}): Promise<MemoryItem> {
  return apiRequest<MemoryItem>("/memory/items", {
    method: "POST",
    auth: true,
    body: input,
  });
}

export async function deleteMemoryItem(id: string): Promise<void> {
  await apiRequest<void>(`/memory/items/${id}`, { method: "DELETE", auth: true });
}

export async function recallMemories(query: string, topK = 6) {
  return apiRequest<
    Array<{
      memory_id: string;
      kind: MemoryKind;
      title: string | null;
      content: string;
      score: number;
    }>
  >("/memory/recall", {
    method: "POST",
    auth: true,
    body: { query, top_k: topK },
  });
}

export async function listGoals(): Promise<CareerGoal[]> {
  return apiRequest<CareerGoal[]>("/memory/goals", { auth: true });
}

export async function createGoal(input: {
  title: string;
  description: string;
  status?: string;
}): Promise<CareerGoal> {
  return apiRequest<CareerGoal>("/memory/goals", {
    method: "POST",
    auth: true,
    body: { status: "active", ...input },
  });
}

export async function listFavorites(): Promise<FavoriteVerse[]> {
  return apiRequest<FavoriteVerse[]>("/memory/favorites", { auth: true });
}

export async function addFavorite(input: {
  verse_id: string;
  note?: string;
}): Promise<FavoriteVerse> {
  return apiRequest<FavoriteVerse>("/memory/favorites", {
    method: "POST",
    auth: true,
    body: input,
  });
}

export async function listJournals(): Promise<JournalEntry[]> {
  return apiRequest<JournalEntry[]>("/memory/journals", { auth: true });
}

export async function createJournal(input: {
  title?: string;
  content: string;
  mood_note?: string;
}): Promise<JournalEntry> {
  return apiRequest<JournalEntry>("/memory/journals", {
    method: "POST",
    auth: true,
    body: input,
  });
}

export async function listReflections(): Promise<ReflectionEntry[]> {
  return apiRequest<ReflectionEntry[]>("/memory/reflections", { auth: true });
}

export async function createReflection(input: {
  journal_id: string;
  content: string;
  verse_id?: string;
}): Promise<ReflectionEntry> {
  return apiRequest<ReflectionEntry>("/memory/reflections", {
    method: "POST",
    auth: true,
    body: input,
  });
}

export async function saveConversationSummary(input: {
  conversation_id?: string;
  title?: string;
  summary: string;
  messages_preview?: string[];
}): Promise<MemoryItem> {
  return apiRequest<MemoryItem>("/memory/conversations/summary", {
    method: "POST",
    auth: true,
    body: input,
  });
}

File: apps/api/app/api/v1/schemas/memory.py
"""Memory API schemas."""

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field

from app.domain.enums.memory_kind import MemoryKind


class MemoryItemResponse(BaseModel):
    """Serialized memory item."""

    id: UUID
    kind: MemoryKind
    title: str | None
    content: str
    source_id: UUID | None
    metadata: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime | None = None
    updated_at: datetime | None = None


class MemoryCreateRequest(BaseModel):
    """Create a free-form memory note."""

    kind: MemoryKind = MemoryKind.NOTE
    title: str | None = Field(default=None, max_length=255)
    content: str = Field(min_length=1, max_length=8000)
    source_id: UUID | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class MemoryRecallRequest(BaseModel):
    """Semantic recall against the user's vector memory."""

    query: str = Field(min_length=1, max_length=2000)
    top_k: int = Field(default=6, ge=1, le=20)
    kinds: list[MemoryKind] | None = None


class RecalledMemoryResponse(BaseModel):
    """One recalled memory hit."""

    memory_id: UUID
    kind: MemoryKind
    title: str | None
    content: str
    score: float
    source_id: UUID | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class CareerGoalCreateRequest(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    description: str = Field(min_length=1, max_length=4000)
    status: str = Field(default="active", pattern="^(active|paused|achieved)$")


class CareerGoalResponse(BaseModel):
    id: UUID
    title: str
    description: str
    status: str
    memory_item_id: UUID | None
    created_at: datetime
    updated_at: datetime


class FavoriteVerseCreateRequest(BaseModel):
    verse_id: UUID
    note: str | None = Field(default=None, max_length=2000)


class FavoriteVerseResponse(BaseModel):
    id: UUID
    verse_id: UUID
    note: str | None
    memory_item_id: UUID | None
    citation: str | None = None
    translation: str | None = None
    created_at: datetime


class JournalCreateRequest(BaseModel):
    title: str | None = Field(default=None, max_length=255)
    content: str = Field(min_length=1, max_length=12000)
    mood_note: str | None = Field(default=None, max_length=255)


class JournalResponse(BaseModel):
    id: UUID
    title: str | None
    content: str
    mood_note: str | None
    created_at: datetime
    updated_at: datetime


class ReflectionCreateRequest(BaseModel):
    journal_id: UUID
    content: str = Field(min_length=1, max_length=8000)
    verse_id: UUID | None = None


class ReflectionResponse(BaseModel):
    id: UUID
    journal_id: UUID
    verse_id: UUID | None
    content: str
    created_at: datetime


class ConversationSummaryRequest(BaseModel):
    """Persist a conversation summary into vector memory."""

    conversation_id: UUID | None = None
    title: str | None = Field(default=None, max_length=255)
    summary: str = Field(min_length=1, max_length=8000)
    messages_preview: list[str] = Field(default_factory=list, max_length=40)
